### DISCUSSION TITLE
Questions about RIS coverage map calculation and dotty behavior

### DISCUSSION THREAD

**Discussion by ekilcioglu** [NVlabs/sionna#514] [Q&A]
Hello,

I'm working on simulating RIS-assisted propagation environments using Sionna RT. After running my simulations and obtaining coverage maps, I've noticed a dotty behavior in the regions where RIS steers its beams. For example, adjacent tiles in the coverage map often show abrupt changes between low power levels (blue squares) and comparably higher power levels (yellowish squares), as you can see this dotty behavior in the RIS coverage gain figure in attached.

I have a few questions regarding this:

1. Does the current implementation in the ray-tracing tool use a coherent or non-coherent approach for computing the electric field and power in the coverage map calculations?

2. What might be causing this dotty behavior in the coverage maps? If the current implementation uses a coherent approach, I thought using a coherent approach while computing the electric field and the power could cause this problem, although it accurately represents the actual behavior of the RIS. The coherent summation (10*log(|∑E_k|^2)) can lead to constructive and destructive interference patterns, which could explain the high and low power levels in adjacent tiles. I have tried to obtain the coverage maps with higher num_sample values such as 1e8, but I encountered the "kernel dies" issue even though the computer specs I'm using are fairly high level. Therefore, increasing the num_sample value could maybe solve this issue, but I cannot do it right now. I have used 1e7 for the figures attached.

3. If the current approach is coherent, would it be possible to modify it to a non-coherent approach (10log(∑|E_k|^2)) to potentially obtain smoother coverage maps?

Any insights or suggestions you could provide would be welcome. I'm happy to provide more details about my simulation setup if needed.

Thank you for your time and assistance!

Best regards

![RIS_coverage_gain](https://github.com/NVlabs/sionna/assets/159263210/dee2971c-4ed7-4a95-a269-2bd216ebbed2)

**Comment by jhoydis** [comment]
Hi @ekilcioglu,

Coverage maps are non-coherent. The "dotty" behavior you observe has two reasons.

1) Only rays that hit the RIS can also hit a point on the coverage map. This number might be fairly small, even for a very large number of rays. You can essentially calculate the percentage of rays that will hit the RIS by computing the solid angle with respect to a sphere around the transmitter that is covered by the RIS.

2) The size of the cells seems rather small. You can simply reduce it to obtain a less noisy result.

Alternatively, you can compute a coverage map by explicitly placing receivers at the centers of the cells, compute the exact channel impulse responses, and then their non-coherent/or coherent energy. This will provide the most accurate result as it also includes diffraction from the RIS' edges. However, depending on the number of grid points, it will also take a while to compute and you might no be able to compute all channels in one go.

What is the height of the transmitter and what is the height of the coverage map in your scenario? You may also not want to place the transmitter at the same height as the coverage map (if this is the case for you).

Hope that helps.

**Comment by ekilcioglu** [comment]
Hello @jhoydis,

First of all, I appreciate your quick and detailed reply about this issue. Then, it is not due to the coherency choice of the calculations. The size of the cells was 0.1 x 0.1 m in these figures. The scenario is an L-shaped corridor scenario with 15 m x 15 m dimensions. I tried it with 0.2 x 0.2 m cell size and the coverage map becomes too blurry this time (attached figure).

I will try to compute a coverage map by placing receivers repeatedly to each center of the cell and compute the power myself (like in the tutorial of the RIS in the documentation when the electric field is calculated in a function), as you have suggested.

The transmitter and the RIS center height are both at 2m while the coverage map height is 1.5 m (default). I include the codes that I'm using to simulate this coverage map as well as the XML file if you want to see them.

Thanks for your ideas once again, it gave me important insights about the structure of the ray-tracing tool.

Best regards


The code:
```
import os
gpu_num = 0 # Use "" to use the CPU
os.environ["CUDA_VISIBLE_DEVICES"] = f"{gpu_num}"
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import sionna
import tensorflow as tf
#tf.random.set_seed(1) # Set global random seed for reproducibility
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
from matplotlib import cm
from sionna import PI
from sionna.rt import load_scene, Transmitter, Receiver, RIS, PlanarArray, r_hat, normalize, Camera

scene = load_scene("L-Shaped_Coverage.xml")
scene.frequency = 5.8e9 # Carrier frequency [Hz]
scene.tx_array = PlanarArray(1,1,0.5,0.5,"iso", "V")
scene.rx_array = PlanarArray(1,1,0.5,0.5,"iso", "V")

# Create the transmitter
tx = Transmitter(name="tx",
                 position=[4.6,14,2])
                #color=[0.0, 0.0, 1.0])
scene.add(tx)

width = 2 # Width [m]
num_rows = num_cols = int(width/(0.5*scene.wavelength))
ris = RIS(name="ris",
          position=[0.2,2,2],
          #color=[0.0, 0.5, 0.5],
          num_rows=num_rows,
          num_cols=num_cols)
scene.add(ris)

# Create the receivers
rx1 = Receiver(name="rx1",
                 position=[10,2,2])
              #color=[0.0, 1.0, 0.0])
scene.add(rx1)

tx.look_at(ris) # Transmitter points towards the RIS

source = tx.position # Location of the origin of the incoming ray
target = rx1.position # Target position
ris.phase_gradient_reflector(source, target)

# Compute coverage map without RIS
cm_no_ris = scene.coverage_map(cm_cell_size=[0.2,0.2],
                               num_samples=1e7,
                               max_depth=5,
                               los=True,
                               reflection=True,
                               diffraction=False,
                               ris=False);
cm_no_ris.show(vmin=-100, show_ris=True, show_rx=True)
plt.title("Coverage without RIS");

# Compute coverage map with RIS
cm_ris = scene.coverage_map(cm_cell_size=[0.2,0.2],
                               num_samples=1e7,
                               max_depth=5,
                               los=True,
                               reflection=True,
                               diffraction=False,
                               ris=True);
cm_ris.show(vmin=-100, show_ris=True, show_rx=True)
plt.title("Coverage with RIS");

RIS_coverage_gain = 10*np.log10(cm_ris._value[0]/cm_no_ris._value[0])

# Visualize the coverage improvements thanks to the RIS
fig = plt.figure()
plt.imshow(RIS_coverage_gain, origin='lower', vmin=0)
plt.colorbar(label='Gain [dB]')
plt.xlabel('Cell index (X-axis)');
plt.ylabel('Cell index (Y-axis)');
plt.title("RIS Coverage Gain");
```

[L-Shaped_Coverage.zip](https://github.com/user-attachments/files/16182256/L-Shaped_Coverage.zip)

![RIS_coverage_gain_v2](https://github.com/NVlabs/sionna/assets/159263210/8edebfd6-1302-40d3-bfa9-0e83a7523c16)

**Comment by jhoydis** [comment]
Glad that it helps.

There might be a related useful feature we could implement in a future release, namely distributing the coverage map computation across multiple runs, e.g., one would define a desired number of samples as well as a "batch_size" (or something similar) which would determine how many rays are shot for each batch. By doing this, one could iteratively compute a coverage map without running into memory issues. Alternatively, one could refine a coverage map by iteratively shooting more rays until a desired precision is reached.

**Comment by ekilcioglu** [comment]
Yes, that could make sense and would be useful for my issue for example! Looking forward to it.

Thanks for your help and have a good evening!
